File: test/java_cfg/edge_cases.java
/*
 * Edge Cases and Boundary Conditions for Java CFG Testing
 * 
 * Purpose: This file contains edge cases, boundary conditions, and potentially
 * problematic scenarios to ensure robust CFG generation. These tests validate
 * the CFG implementation's resilience to unusual code patterns.
 * 
 * CFG Testing Strategy:
 * - Test empty constructs and minimal cases
 * - Validate handling of unreachable code
 * - Test complex expressions and conditions
 * - Ensure graceful handling of unusual syntax
 * - Test boundary conditions for nested structures
 * - Validate post-processing doesn't break edge cases
 * 
 * Edge Case Categories:
 * - Empty constructs (empty blocks, empty loops)
 * - Unreachable code (after returns, breaks)
 * - Complex expressions (compound conditions, method chains)
 * - Unusual control flow patterns
 * - Boundary nesting conditions
 * - Degenerate cases
 * 
 * Expected Behavior:
 * - Empty constructs should create placeholder nodes if needed
 * - Unreachable code should still be modeled but not connected
 * - Complex expressions should be parsed correctly
 * - No crashes or assertion failures on unusual input
 * - CFG structure remains valid even for edge cases
 * 
 * Error Conditions:
 * - Code that may not compile should still generate reasonable CFGs
 * - Parser should handle incomplete constructs gracefully
 * - Post-processing should preserve CFG integrity
 * 
 * Created: 2025-08-14
 * Test Category: Edge Cases and Boundary Conditions
 */

public class EdgeCases {
    
    /**
     * Test 1: Empty Blocks
     * Expected CFG: Placeholder nodes for empty constructs
     * Edge Case: Empty compound statements should not break CFG structure
     * Variables: x (def), empty blocks should not affect variable tracking
     */
    public static void emptyBlocks() {
        int x = 5;
        if (x > 0) {
            // Empty if block
        } else {
            // Empty else block  
        }
        
        for (int i = 0; i < 10; i++) {
            // Empty for loop body
        }
        
        while (x > 0) {
            // Empty while body - this would be infinite loop, but we're testing CFG structure
            break; // Need this to avoid infinite loop
        }
        
        switch (x) {
            case 5:
                // Empty case
                break;
            default:
                // Empty default
        }
    }
    
    /**
     * Test 2: Unreachable Code After Return
     * Expected CFG: Return connects to exit, subsequent statements unreachable
     * Edge Case: Code after return should still be modeled but not connected
     * Variables: x (param), unreachable code should still be parsed
     */
    public static int unreachableAfterReturn(int x) {
        if (x > 0) {
            return x;
            System.out.println("Unreachable after return");  // Unreachable
            x = x + 1;  // Also unreachable
        }
        return -1;
    }
    
    /**
     * Test 3: Unreachable Code After Break
     * Expected CFG: Break connects to loop exit, subsequent statements unreachable
     * Edge Case: Code after break in loop should be modeled but not connected
     * Variables: i (def), unreachable code in loop
     */
    public static void unreachableAfterBreak() {
        for (int i = 0; i < 10; i++) {
            if (i == 5) {
                break;
                System.out.println("Unreachable after break");  // Unreachable
                i++;  // Also unreachable
            }
            System.out.println("Reachable: " + i);
        }
    }
    
    /**
     * Test 4: Complex Conditional Expressions
     * Expected CFG: Condition node with complex expression text
     * Edge Case: Complex boolean expressions should be handled correctly
     * Variables: a, b, c (def), complex condition uses all variables
     */
    public static void complexConditionalExpressions() {
        int a = 5, b = 10, c = 15;
        
        // Complex condition with multiple operators
        if ((a > 0 && b < 20) || (c % 3 == 0 && Math.abs(a - b) > 2)) {
            System.out.println("Complex condition true");
        }
        
        // Nested ternary operators
        int result = a > b ? (c > a ? c : a) : (c > b ? c : b);
        System.out.println("Result: " + result);
    }
    
    /**
     * Test 5: Empty For Loop Components
     * Expected CFG: Graceful handling of missing init, condition, or update
     * Edge Case: For loops can have empty initialization, condition, or update
     * Variables: counter (def), loop with minimal components
     */
    public static void emptyForLoopComponents() {
        int counter = 0;
        
        // For loop with empty initialization (using existing variable)
        for (; counter < 5; counter++) {
            System.out.println("Counter: " + counter);
        }
        
        // For loop with no condition (infinite loop, but we'll break)
        for (int i = 0; ; i++) {
            if (i >= 3) {
                break;
            }
            System.out.println("Infinite loop iteration: " + i);
        }
        
        // For loop with no update (manual update in body)
        for (int j = 0; j < 3; ) {
            System.out.println("Manual update: " + j);
            j++;
        }
    }
    
    /**
     * Test 6: Switch with No Default
     * Expected CFG: Switch with cases but no default case
     * Edge Case: Switch statements don't require default cases
     * Variables: value (def), switch without default
     */
    public static void switchWithoutDefault() {
        int value = 5;
        int result = 0;
        
        switch (value) {
            case 1:
                result = 10;
                break;
            case 2:
                result = 20;
                break;
            case 3:
                result = 30;
                break;
            // No default case - control flows to after switch
        }
        
        System.out.println("Result: " + result);
    }
    
    /**
     * Test 7: Switch with Only Default
     * Expected CFG: Switch with only default case
     * Edge Case: Switch can have only default case
     * Variables: x (def), switch with only default
     */
    public static void switchWithOnlyDefault() {
        int x = 10;
        
        switch (x) {
            default:
                System.out.println("Only default case: " + x);
        }
    }
    
    /**
     * Test 8: Deeply Nested Empty Structures
     * Expected CFG: Multiple levels of empty nesting
     * Edge Case: Maximum nesting depth with empty constructs
     * Variables: flag (def), deep nesting with minimal content
     */
    public static void deeplyNestedEmptyStructures() {
        boolean flag = true;
        
        if (flag) {
            for (int i = 0; i < 1; i++) {
                if (i == 0) {
                    while (flag) {
                        if (true) {
                            for (int j = 0; j < 1; j++) {
                                // Deep nesting with minimal logic
                                flag = false;
                            }
                        }
                    }
                }
            }
        }
    }
    
    /**
     * Test 9: Method Call Chains
     * Expected CFG: Method calls should be captured in node metadata
     * Edge Case: Complex method call chains and expressions
     * Variables: str (def), complex method chaining
     */
    public static void methodCallChains() {
        String str = "Hello World";
        
        // Method call chain
        int length = str.toLowerCase().trim().replace("world", "Java").length();
        
        // Nested method calls in expressions
        boolean result = Math.abs(Math.max(5, 10) - Math.min(3, 7)) > 0;
        
        System.out.println("Length: " + length + ", Result: " + result);
    }
    
    /**
     * Test 10: Try-Catch with Empty Blocks
     * Expected CFG: Exception handling with minimal content
     * Edge Case: Empty try/catch/finally blocks
     * Variables: x (def), minimal exception handling
     */
    public static void tryCatchEmptyBlocks() {
        int x = 5;
        
        try {
            // Empty try block
        } catch (Exception e) {
            // Empty catch block
        } finally {
            // Empty finally block
        }
        
        // Try with only finally
        try {
            x = x / 1;  // Safe operation
        } finally {
            System.out.println("Finally executed");
        }
    }
    
    /**
     * Test 11: Multiple Consecutive Jumps
     * Expected CFG: Multiple jump statements in sequence
     * Edge Case: Consecutive returns/breaks (though only first is reachable)
     * Variables: x (param), multiple jump paths
     */
    public static int multipleConsecutiveJumps(int x) {
        if (x > 10) {
            return x;
            return x + 1;  // Unreachable
            return x + 2;  // Also unreachable
        }
        
        for (int i = 0; i < 10; i++) {
            if (i == 3) {
                break;
                continue;  // Unreachable
                break;     // Also unreachable
            }
        }
        
        return 0;
    }
    
    /**
     * Test 12: Enhanced For with Complex Iterables
     * Expected CFG: Enhanced for loop with method call as iterable
     * Edge Case: Complex expressions as iterable source
     * Variables: line (def in loop), complex iterable expression
     */
    public static void enhancedForComplexIterables() {
        // Enhanced for with method call returning iterable
        for (String line : java.util.Arrays.asList("a", "b", "c")) {
            System.out.println("Line: " + line);
        }
        
        // Enhanced for with array expression
        int[] numbers = {1, 2, 3};
        for (int num : numbers.length > 0 ? numbers : new int[]{0}) {
            System.out.println("Number: " + num);
        }
    }
    
    /**
     * Test 13: Assert Statements
     * Expected CFG: Assert statements as conditional constructs
     * Edge Case: Assert statements may be enabled/disabled
     * Variables: x (def), assert conditions
     */
    public static void assertStatements() {
        int x = 5;
        
        // Simple assert
        assert x > 0;
        
        // Assert with message
        assert x < 100 : "x should be less than 100, got: " + x;
        
        // Assert in complex expression
        assert (x % 2 == 1) || (x % 3 == 2) : "Complex assertion failed";
    }
    
    /**
     * Test 14: Synchronized Blocks
     * Expected CFG: Synchronized blocks as structured constructs
     * Edge Case: Concurrency constructs with lock acquisition
     * Variables: obj (def), synchronized block content
     */
    public static void synchronizedBlocks() {
        Object obj = new Object();
        int counter = 0;
        
        synchronized (obj) {
            counter++;
            System.out.println("In synchronized block: " + counter);
        }
        
        // Nested synchronized blocks
        synchronized (obj) {
            synchronized (obj) {
                counter += 2;
                System.out.println("Nested synchronized: " + counter);
            }
        }
    }
}

File: test/c_cfg/edge_cases.c
/**
 * Edge cases and boundary conditions for CFG testing
 * 
 * This file contains edge cases, boundary conditions, and potential error scenarios:
 * - Empty functions and blocks
 * - Single statement blocks (no braces)
 * - Switch statements without default cases
 * - Multiple consecutive jump statements
 * - Unreachable code scenarios
 * - Complex conditional expressions
 * - Very long switch statements
 * - Functions with many parameters
 * 
 * Use for: Robustness testing, error condition handling
 * 
 * These functions test the CFG implementation's ability to handle unusual
 * or extreme cases gracefully, including boundary conditions that might
 * cause parsing errors or generate unexpected CFG structures.
 */

// Empty function
void empty_function() {
}

// Function with only return
int only_return() {
    return 42;
}

// Empty blocks
int empty_blocks(int x) {
    if (x > 0) {
        // Empty if block
    }
    
    while (x > 10) {
        // Empty while block
    }
    
    for (int i = 0; i < x; i++) {
        // Empty for block
    }
    
    switch (x) {
        case 1:
            // Empty case
            break;
        default:
            // Empty default
            break;
    }
    
    return x;
}

// Single statement blocks
int single_statement_blocks(int x) {
    if (x > 0)
        x = x + 1;  // No braces
    
    while (x < 100)
        x = x * 2;  // No braces
    
    for (int i = 0; i < 5; i++)
        x = x + i;  // No braces
    
    return x;
}

// Switch without default
int switch_no_default(int x) {
    int result = x;
    switch (x) {
        case 1:
            result = 10;
            break;
        case 2:
            result = 20;
            break;
    }  // No default case
    return result;
}

// Switch with only default
int switch_only_default(int x) {
    int result = x;
    switch (x) {
        default:
            result = -1;
    }
    return result;
}

// All cases fall through
int switch_all_fallthrough(int x) {
    int result = 0;
    switch (x) {
        case 1:
            result = result + 1;
        case 2:
            result = result + 2;
        case 3:
            result = result + 3;
        default:
            result = result + 10;
    }
    return result;
}

// Multiple consecutive breaks/continues
int multiple_breaks_continues(int limit) {
    int sum = 0;
    for (int i = 0; i < limit; i++) {
        if (i == 5) {
            continue;
            continue;  // Unreachable
        }
        
        if (i == 10) {
            break;
            break;  // Unreachable
        }
        
        sum = sum + i;
    }
    return sum;
}

// Multiple consecutive returns
int multiple_consecutive_returns(int x) {
    if (x > 0) {
        return x;
        return -x;  // Unreachable
    }
    return 0;
}

// Goto jumping over declarations
int goto_over_declarations(int x) {
    if (x < 0) {
        goto skip_declarations;
    }
    
    int a = 5;
    int b = 10;
    x = a + b;
    
skip_declarations:
    return x;
}

// Backward goto creating loop
int backward_goto_loop(int n) {
    int sum = 0;
    
loop_start:
    if (n <= 0) {
        return sum;
    }
    
    sum = sum + n;
    n = n - 1;
    goto loop_start;
}

// Multiple labels
int multiple_labels(int x) {
label1:
label2:
label3:
    x = x + 1;
    
    if (x == 1) {
        goto label1;
    }
    if (x == 2) {
        goto label2;
    }
    if (x == 3) {
        goto label3;
    }
    
    return x;
}

// Complex expressions in conditions
int complex_expressions(int x, int y, int z) {
    // Complex if condition
    if ((x > 0 && y < 10) || (z == 5 && x != y)) {
        x = x + y + z;
    }
    
    // Complex while condition  
    while ((x > 0 && x < 100) && (y > 0 || z < 0)) {
        if (x % 2 == 0 && y % 3 == 0 && z % 5 == 0) {
            break;
        }
        x = x - 1;
        y = y + 1;
        z = z * 2;
    }
    
    // Complex for condition
    for (int i = 0; (i < x && i < y) || i < z; i++) {
        if (i * i > x + y + z) {
            continue;
        }
        x = x + i;
    }
    
    return x + y + z;
}

// Nested ternary operators (not control flow, but complex)
int ternary_expressions(int x, int y, int z) {
    int result = (x > y) ? (y > z ? x : z) : (x > z ? y : x);
    return result;
}

// Very long switch statement
int long_switch(int x) {
    int result = 0;
    switch (x) {
        case 1: result = 1; break;
        case 2: result = 4; break;
        case 3: result = 9; break;
        case 4: result = 16; break;
        case 5: result = 25; break;
        case 6: result = 36; break;
        case 7: result = 49; break;
        case 8: result = 64; break;
        case 9: result = 81; break;
        case 10: result = 100; break;
        case 11: result = 121; break;
        case 12: result = 144; break;
        case 13: result = 169; break;
        case 14: result = 196; break;
        case 15: result = 225; break;
        default: result = x * x; break;
    }
    return result;
}

// Function with many parameters (tests entry node)
int many_parameters(int a, int b, int c, int d, int e, 
                   int f, int g, int h, int i, int j) {
    return a + b + c + d + e + f + g + h + i + j;
}

// Minimal valid C program
int minimal() {
    return 0;
}